File: backend/src/schemas/umap.py
from __future__ import annotations
from pydantic import BaseModel


class UmapPoint2D(BaseModel):
    song_id: str
    x: float
    y: float
    title: str | None = None
    artist: str | None = None


class UmapResponse(BaseModel):
    points_2d: list[UmapPoint2D]
    features_used: list[str]

File: frontend/src/types/umap.ts
export interface UmapPoint2D {
  song_id: string;
  x: number;
  y: number;
  title: string | null;
  artist: string | null;
  /** Nearest neighbours in the high-dimensional feature space, closest first. */
  neighbors: string[];
}

export interface UmapResponse {
  points_2d: UmapPoint2D[];
  features_used: string[];
}
